core/include/detray/utils/root_finding.hpp

```diff
@@ -0,0 +1,302 @@
+/** Detray library, part of the ACTS project (R&D line)
+ *
+ * (c) 2024 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#pragma once
+
+// Project include(s).
+#include "detray/definitions/detail/algebra.hpp"
+#include "detray/definitions/detail/math.hpp"
+#include "detray/definitions/detail/qualifiers.hpp"
+#include "detray/definitions/units.hpp"
+#include "detray/utils/invalid_values.hpp"
+
+// System include(s).
+#include <algorithm>
+#include <array>
+#include <iostream>
+#include <limits>
+#include <stdexcept>
+#include <string>
+#include <tuple>
+
+namespace detray {
+
+/// @brief Try to find a bracket around a root
+///
+/// @param [in] a lower initial boundary
+/// @param [in] b upper initial boundary
+/// @param [in] f function for which to find the root
+/// @param [out] bracket bracket around the root
+/// @param [in] k scale factor with which to widen the bracket at every step
+///
+/// @see Numerical Recepies pp. 445
+///
+/// @return whether a bracket was found
+template <typename scalar_t, typename function_t>
+DETRAY_HOST_DEVICE inline bool expand_bracket(const scalar_t a,
+                                              const scalar_t b, function_t &f,
+                                              std::array<scalar_t, 2> &bracket,
+                                              const scalar_t k = 1.f) {
+
+    if (a == b) {
+        throw std::invalid_argument(
+            "Root bracketing: Not a valid start interval [" +
+            std::to_string(a) + ", " + std::to_string(b) + "]");
+    }
+
+    scalar_t lower{a > b ? b : a};
+    scalar_t upper{a > b ? a : b};
```

Comment:
This differs from the _Numerical Recipes_ definitions as $l$ is the value with the lower value $x$ rather than the lower value of $f(x)$. Is that intended?

Reply:
Yes, I found an ordered interval easier to reason about during debugging. So, the bracketing function returns an ordered bracket now, which at a later step gets put into ```a``` and ```b``` the according to function value, the way it is done in Numerical Recipes. I think, because the update of the bracket in the hot loop is more efficient this way?

Reply:
Alright, that makes sense. Just checking. :slightly_smiling_face: